inform7/Internal/Inter/Architecture32Kit/Sections/Glk.i6t

```diff
@@ -4,1001 +4,499 @@ Glk is a fully-featured input-output model, designed by Andrew Plotkin for use
+	@push initlen;
+	@push maxlen;
+	@push buf;
+	@push win;
+	@glk 208 4 0;
+	win_obj = FindGlkWindowFromRefNum(win);
+	if (win_obj) {
+		win_obj.text_input_status = INPUT_STATUS_ACTIVE_LINE;
+		win_obj.line_input_buffer_addr = buf;
+		win_obj.line_input_buffer_maxlen = maxlen;
+		win_obj.line_input_buffer_uni = false;
+	}
+	return 0;
+];
+
+[ glk_request_line_event_uni win buf maxlen initlen win_obj;
+	@push initlen;
+	@push maxlen;
+	@push buf;
+	@push win;
+	@glk 321 4 0;
+	win_obj = FindGlkWindowFromRefNum(win);
+	if (win_obj) {
+		win_obj.text_input_status = INPUT_STATUS_ACTIVE_LINE_UNI;
+		win_obj.line_input_buffer_addr = buf;
+		win_obj.line_input_buffer_maxlen = maxlen;
+		win_obj.line_input_buffer_uni = true;
+	}
+	return 0;
 ];
 
-[ glk_date_to_time_utc _vararg_count;
-  ! glk_date_to_time_utc(&{int, int, int, int, int, int, int, int}, &{int, uint, int})
-  @glk 364 _vararg_count 0;
-  return 0;
+[ glk_request_mouse_event win win_obj;
+	@push win;
+	@glk 212 1 0;
+	win_obj = FindGlkWindowFromRefNum(win);
+	if (win_obj) {
+		win_obj.requesting_mouse = true;
+	}
+	return 0;
 ];
 
-[ glk_date_to_time_local _vararg_count;
-  ! glk_date_to_time_local(&{int, int, int, int, int, int, int, int}, &{int, uint, int})
-  @glk 365 _vararg_count 0;
-  return 0;
-];
+@h Suspending and resuming text input.
+These functions allow the author to suspend and then resume a window's text
+input requests.
 
-[ glk_date_to_simple_time_utc _vararg_count ret;
-  ! glk_date_to_simple_time_utc(&{int, int, int, int, int, int, int, int}, uint) => int
-  @glk 366 _vararg_count ret;
-  return ret;
+=
+[ SuspendTextInput win no_input_echoing;
+	if (win.text_input_status == INPUT_STATUS_ACTIVE_CHAR or INPUT_STATUS_ACTIVE_CHAR_UNI) {
+		glk_cancel_char_event(win.glk_ref);
+	}
+	else if (win.text_input_status == INPUT_STATUS_ACTIVE_LINE or INPUT_STATUS_ACTIVE_LINE_UNI) {
+		glk_cancel_line_event(win.glk_ref, gg_arguments);
+		! Manually echo the command if required
+		if (BasicInformKit`MANUAL_INPUT_ECHOING_CFGF && Cached_Glk_Gestalts-->gestalt_LineInputEcho && no_input_echoing == 0) {
+			glk_set_style(style_Input);
+			print (TEXT_TY_Say) win.current_line_input;
+			glk_set_style(style_Normal);
+			glk_put_char(10); ! newline
+		}
+	}
 ];
 
-[ glk_date_to_simple_time_local _vararg_count ret;
-  ! glk_date_to_simple_time_local(&{int, int, int, int, int, int, int, int}, uint) => int
-  @glk 367 _vararg_count ret;
-  return ret;
+[ ResumeTextInput win;
+	switch (win.text_input_status) {
+		INPUT_STATUS_SUSPENDED_CHAR:
+			glk_request_char_event(win.glk_ref);
+		INPUT_STATUS_SUSPENDED_CHAR_UNI:
+			glk_request_char_event_uni(win.glk_ref);
+		INPUT_STATUS_SUSPENDED_LINE:
+			glk_request_line_event(win.glk_ref, win.line_input_buffer_addr, win.line_input_buffer_maxlen, CopyLineInputTextToBuffer(win));
+		INPUT_STATUS_SUSPENDED_LINE_UNI:
+			glk_request_line_event_uni(win.glk_ref, win.line_input_buffer_addr, win.line_input_buffer_maxlen, CopyLineInputTextToBuffer(win));
+	}
```

Comment:
Should we add a RTP if you try to resume and the current status isn't suspended?

Reply:
Probably not a bad idea.  In my original version I was assuming that it would most likely crash if you did that anyway, since it would be trying to write to memory that might not be valid any more, but that could equally just trash some memory without crashing or show no apparent ill effects.  Doesn't hurt to have an explicit check, though ideally only very-well-tested core extensions would be doing these things anyway.

Reply:
I'm not sure if it should be considered a big though. For example, there could be some code which wants to repeatedly suspend and resume even though sometimes the game might be waiting for keyboard input and other times only waiting for hyperlinks or timers. In that case it would be harmless to have nothing to resume.

But on the other hand, such code could easily check whether keyboard input is active before suspending/resuming. So it might be better to raise a RTP by default because it's easy enough to avoid if keyboard input isn't active.

Reply:
I think attempting to cancel when there is nothing pending should be ignored without error.  Attempting to resume or replace when there's nothing suspended seems more problematic to me.

Suspending when there is already something suspended is another interesting case.  I definitely had it in mind that it should be possible to suspend line input, print some text, then perform character input (typically just a "press key to continue" or perhaps a confirmation prompt), print some more text, and then finally resume the original line input.  I'm less sure if it's reasonable to allow nested line inputs (my extension didn't support that) -- while it might be nice if it could stack them, it's likely both would have used the same buffer anyway and wouldn't coexist nicely.

In terms of hyperlinks specifically, while I think the player would normally only click hyperlinks while the game is paused waiting for input (usually line input, but possibly character input), it's certainly possible that they might click a link while the story is in the midst of printing output or waiting on a timer.

Reply:
I did implement an input state stack, but decided to remove it from this PR. I'll probably revisit it in an extension.